_posts/2024-03-30-nsb-asb-retries.md

```diff
@@ -0,0 +1,63 @@
+
+When you let NServiceBus create your ASB queues, this value is set to `int.MaxValue` to prevent it from interfering with the NServiceBus recoverability configuration. You can take control of the queue creation yourself and set a more restrictive message processing limit on the queues. This is a simple solution but has some disadvantages:
+
+* NServiceBus doesn't allow the configuration of this value, not via [code-first configuration](https://docs.particular.net/transports/azure-service-bus/configuration) nor the [CLI tool](https://docs.particular.net/transports/azure-service-bus/operational-scripting). Therefore you need to create the queues with all the other required settings manually which can be error-prone.
+* The message is moved to the Azure Service Bus dead-letter queue instead of the NServiceBus configured error queue which makes it more difficult to integrate with ServiceControl.
+
+### Shortcutting the pipeline
+
+You can read the delivery count value of the incoming message within the pipeline and decide to invoke the recoverability pipeline directly, preventing the failing message handler code from being invoked at all. The following code snippets show how to do that:
+
+```csharp
+class InterceptingBehavior : Behavior<ITransportReceiveContext>
+{
+    public override Task Invoke(ITransportReceiveContext context, Func<Task> next)
+    {
+        // retrieve the ASB native message type to access additional message metadata
+        var asbMessage = context.Extensions.Get<ServiceBusReceivedMessage>();
+        if (asbMessage.DeliveryCount >= 10)
+        {
+            throw new MessageDeliveriesExceededException();
+        }
+
+        return next();
+    }
+}
+
+class MessageDeliveriesExceededException : Exception {}
+```
+
+The custom behavior will inspect the delivery count tracked by ASB and throw a custom exception if it exceeds our defined threshold. The behavior and exception need to be registered during endpoint configuration:
+
+```csharp
+configuration.Pipeline.Register(new InterceptingBehavior(), "Prevents endless retries due to process crashes");
+// unrecoverable exceptions will move messages directly to the error queue, bypassing regular retries
+configuration.Recoverability().AddUnrecoverableException<MessageDeliveriesExceededException>();
+```
+
+## Conclusion
+
+It can be helpful to understand how NServiceBus' recoverability feature works to understand its limitations. Error cases that cause a total process shutdown should be very rare and for most users it's probably more useful to have good monitoring of your endpoint health to detect endpoint crashes rather than applying the described solutions. However, there might be cases where the described solutions can help to mitigate the problem more reliably and it's once more a good example of how powerful custom pipeline behaviors are.
```

Comment:
```suggestion
Understanding how NServiceBus' recoverability feature works can help understand its limitations. Error cases that cause a total process shutdown should be very rare. For most users, it's probably more helpful to monitor their endpoint health well to detect endpoint crashes rather than apply the described solutions. However, there might be cases where the described solutions can help mitigate the problem more reliably, and it's once more a good example of how powerful custom pipeline behaviors are.
```